text/0000-pi-types.md

```diff
@@ -0,0 +1,306 @@
+2. Monomorphisation-time errors, i.e. errors that happens during codegen of
+   generic functions. We try to avoid adding _more_ of these (as noted by
+   petrochenkov, these [already exists](https://github.com/rust-lang/rfcs/pull/1657#discussion_r68202733))
+
+# Detailed design
+[design]: #detailed-design
+
+## Notation
+
+We make use of [the standard type judgement notation](https://en.wikipedia.org/wiki/Type_rule).
+We also make use of the `𝓤` notation for (some) type universe.
+
+## The new value-type construct, `const`
+
+Declaring a parameter `const x: T` allows using `x` in both an expression context
+(as a value of type `T`) and a type context (as a type parameter). In a sense,
+const "bridges" the world between values and types, since it allows us to
+declare value dependent types ([`ε → τ` constructors](https://en.wikipedia.org/wiki/Dependent_type)).
+
+Such a parameter is declared, like type parameters, in angle brackets (e.g.
+`struct MyStruct<const x: usize>`).
+
+The expr behavior is described as:
+
+    ValueParameterDeclaration:
+      Γ ⊢ const x: T
+      ──────────────
+      Γ ⊢ x: T
+
+In human language, this simply means that one can use a constant parameter,
+`const x: T`, in expression context, as a value of type `T`.
+
+On the type level, we use the very same semantics as the ones generic
+parameters currently follows.
+
+## `const fn`s as Π-constructors
+
+We are interested in value dependency, but at the same time, we want to avoid
+complications such as [SMT-solvers](https://en.wikipedia.org/wiki/Satisfiability_modulo_theories).
+
+We achieve this by `const fn`, which allows us to take some const parameter and
+map it by some arbitrary, pure function, following the rules described in [RFC
+0911](https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md#detailed-design).
+
+## Type inference
+
+Since we are able to evaluate the function at compile time, we can easily infer
+const parameters, by adding an unification relation, simply
+
+    PiRelationInference
+      Γ ⊢ y = f(x)
+      Γ ⊢ T: U y
+      ──────────────
+      Γ ⊢ T: U f(x)
+
+Informally, this means that, you can substitute equal terms (in this case, `const fn` relations).
+
+The relational edge between two const parameters is simple a const fn, which is
+resolved under unification.
+
+We add an extra rule to improve inference:
+
+    DownLiftEquality:
+      Γ ⊢ T: A → 𝓤
+      Γ ⊢ c: A
+      Γ ⊢ x: A
+      Γ ⊢ a: T c
+      Γ ⊢ a: T x
+      ────────────
+      Γ ⊢ c = x
+
+So, if two types share constructor by some Π-constructor, share a value, their
+value parameter is equal. Take `a: [u8; 4]` as an example. If we have some
+unknown variable `x`, such that `a: [u8; x]`, we can infer that `x = 4`.
+
+This allows us to infer e.g. array length parameters in functions:
+
+```rust
+// [T; N] is a constructor, T → usize → 𝓤 (parameterize over T and you get A → 𝓤).
+fn foo<const n: usize, const l: [u32; n]>() -> [u32; n] {
```

Comment:
In my browser, the letter `lowercase L (l)` and the number `1` are almost identical. For a little bit I actually thought it _was_ the number `1` and I was really confused. Something like `arr` instead of `l` would eliminate any possible confusion